File: observability/health.go
package observability

import (
	"encoding/json"
	"net/http"
	"time"

	"gorm.io/gorm"
)

type HealthHandler struct {
	db *gorm.DB
}

func NewHealthHandler(db *gorm.DB) *HealthHandler {
	return &HealthHandler{db: db}
}

func (h *HealthHandler) Live(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	_ = json.NewEncoder(w).Encode(map[string]interface{}{
		"status":    "UP",
		"service":   "doctor-service",
		"timestamp": time.Now().Format(time.RFC3339),
	})
}

func (h *HealthHandler) Ready(w http.ResponseWriter, r *http.Request) {
	dbStatus := "UP"
	httpStatus := http.StatusOK

	if h.db != nil {
		sqlDB, err := h.db.DB()
		if err != nil || sqlDB.Ping() != nil {
			dbStatus = "DOWN"
			httpStatus = http.StatusServiceUnavailable
		}
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(httpStatus)
	_ = json.NewEncoder(w).Encode(map[string]interface{}{
		"status":    dbStatus,
		"service":   "doctor-service",
		"database":  dbStatus,
		"timestamp": time.Now().Format(time.RFC3339),
	})
}

File: python_backend/observability/health.py
from datetime import datetime

class HealthHandler:
    def __init__(self, db=None):
        self.db = db

    def live(self, request):
        return {
            "status": "UP",
            "service": "doctor-service",
            "timestamp": datetime.now().isoformat()
        }, 200

    def ready(self, request):
        db_status = "UP"
        http_status = 200

        if self.db is not None:
            # Assuming self.db has a ping or similar method in Python
            try:
                # self.db.ping()
                pass
            except Exception:
                db_status = "DOWN"
                http_status = 503

        return {
            "status": db_status,
            "service": "doctor-service",
            "database": db_status,
            "timestamp": datetime.now().isoformat()
        }, http_status
